server/index.js

```diff
@@ -447,30 +435,12 @@ io.on('connection', function(socket) {
             return;
         }
 
-        runAndCatchErrors(game, () => {           
-            game.playerLeave(socket.request.user.username, 'has left the game');
-
-            var player = game.players[socket.request.user.username];
-
-            if(!player) {
-                return;
-            }
-
-            if(game.isSpectator(player)) {
-                delete game.players[socket.request.user.username];
-                io.to(player.id).emit('gamestate', game.getState(player.name));
-            } else {
-                game.players[socket.request.user.username].left = true;
-
-                if(game.started && !game.finishedAt) {
-                    game.finishedAt = new Date();
-                    game.saveGame();
-                }                
-            }
+        runAndCatchErrors(game, () => {
+            game.leave(socket.request.user.username);
+            // Force update of game state to make sure spectators get disconnected.
+            socket.emit('gamestate', game.getState(socket.request.user.username));
```

Comment:
Just tested, this line is explicitly needed for spectators. This is because we delete the spectator from the players array, so when it updates the game state for everyone in the list they won't receive it (since they're not in the list). Without some sort of explicit update, the "Leave" button for spectators doesn't revert the UI back to the game list.